---
title: "MutationalPatterns"
date: "`r Sys.time()`"
output: rmarkdown::html_document
resource_files:
  - outputs/sig_plots/Sig
  - outputs/sig_plots/SBS
  - outputs/sig_plots/DBS
  - outputs/sig_plots/ID
---

```{r knitr_opts, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup, message=FALSE}
require(sigrap)
require(gpgr)
require(MutationalPatterns)
ref_genome <- "BSgenome.Hsapiens.UCSC.hg38"
require(ref_genome, character.only = TRUE)
require(devtools)
require(dplyr)
require(patchwork)

# Check for optional TxDb package needed for strand bias analysis
txdb_available <- requireNamespace("TxDb.Hsapiens.UCSC.hg38.knownGene", quietly = TRUE)
if (txdb_available) {
  message("✓ TxDb.Hsapiens.UCSC.hg38.knownGene available - strand bias analysis enabled")
} else {
  message("ℹ Install TxDb.Hsapiens.UCSC.hg38.knownGene for strand bias analysis:")
  message("  BiocManager::install('TxDb.Hsapiens.UCSC.hg38.knownGene')")
}
```

## Data Preparation

```{r data_prep}
params <- list(
  vcf = system.file("extdata/umccrise/snv/somatic-ensemble-PASS.vcf.gz", package = "gpgr"),
  nm = "sampleA",
  outdir = "outputs",
  include_rainfall = FALSE,
  include_strand_bias = FALSE
)

gr <- MutationalPatterns::read_vcfs_as_granges(
  vcf_files = params$vcf,
  sample_names = params$nm,
  genome = ref_genome,
  group = "auto+sex",
  type = "all"
)
```

## SNVs (SBS)

### Plots

```{r snv_plots, fig.width=12, fig.height=12, out.width="90%"}
snv_counts <- sigrap::sig_count_snv(vcf_gr = gr, ref_genome = ref_genome)
p_snvs <- sigrap::sig_plot_snv(
  vcf_gr = gr, snv_counts = snv_counts$snv_counts,
  ref_genome = ref_genome, rainfall = params$include_rainfall
)
p_snvs$p_river + p_snvs$p_heatmap + p_snvs$p_spectrum + p_snvs$p_96_profile + patchwork::plot_layout(ncol = 1)
```

```{r rainfall, out.width="100%", out.height="600px", fig.width=20, fig.height=8, eval=params$include_rainfall}
if (!is.null(p_snvs$p_rainfall)) {
  p_snvs$p_rainfall
}
```

### Strand Bias Analysis

```{r strand_bias, fig.width=16, fig.height=10, out.width="100%", eval=params$include_strand_bias}
# Optional strand bias analysis
strand_plots <- sigrap::sig_plot_strand_bias(vcf_gr = gr, ref_genome = ref_genome)

# Display plots in 2x2 layout
(strand_plots$p_transcriptional_bias / strand_plots$p_transcriptional_effect) | 
(strand_plots$p_replicative_bias / strand_plots$p_replicative_effect)
```

### Signature Contributions

```{r snv_contributions}
sigs_snv_2015 <-
  sigrap::cosmic_signatures_2015 |>
  {
    \(sigs) sigrap::sig_contribution(mut_mat = snv_counts$snv_counts, signatures = sigs)
  }()

sigs_snv_2020 <-
  MutationalPatterns::get_known_signatures(
    muttype = "snv",
    incl_poss_artifacts = TRUE
  ) |>
  {
    \(sigs) sigrap::sig_contribution(mut_mat = snv_counts$snv_counts, signatures = sigs)
  }()

sigs_snv_2015 |>
  sigrap::sig_contribution_table(type = "Sig", outdir = params$outdir) |>
  knitr::kable(format = "html") |>
  kableExtra::kable_styling(c("hover", "striped"), font_size = 12) |>
  kableExtra::scroll_box(height = "400px")

sigs_snv_2020 |>
  sigrap::sig_contribution_table(type = "SBS", outdir = params$outdir) |>
  knitr::kable(format = "html") |>
  kableExtra::kable_styling(c("hover", "striped"), font_size = 12) |>
  kableExtra::scroll_box(height = "400px")
```

## DBS

### Plots

```{r dbs_plots, fig.width=12, fig.height=10, out.width="90%"}
dbs_counts <- sigrap::sig_count_dbs(vcf_gr = gr)
p_dbs <- sigrap::sig_plot_dbs(dbs_counts = dbs_counts)
p_dbs$p_dbs_main / p_dbs$p_dbs_cont
```

### Signature Contributions

```{r dbs_contributions}
sigs_dbs <-
  MutationalPatterns::get_known_signatures(muttype = "dbs") |>
  {
    \(sigs) sigrap::sig_contribution(mut_mat = dbs_counts, signatures = sigs)
  }()
sigs_dbs |>
  sigrap::sig_contribution_table(type = "DBS", outdir = params$outdir) |>
  knitr::kable(format = "html") |>
  kableExtra::kable_styling(c("hover", "striped"), font_size = 12) |>
  kableExtra::scroll_box(height = "400px")
```

## INDELs

### Plots

```{r indel_plots, fig.width=12, fig.height=10, out.width="90%"}
indel_counts <- sigrap::sig_count_indel(vcf_gr = gr, ref_genome = ref_genome)
p_indels <- sigrap::sig_plot_indel(indel_counts = indel_counts)
p_indels$p_indel_main / p_indels$p_indel_cont
```

### Signature Contributions

```{r indel_contributions}
sigs_indel <-
  MutationalPatterns::get_known_signatures(muttype = "indel") |>
  {
    \(sigs) sigrap::sig_contribution(mut_mat = indel_counts, signatures = sigs)
  }()
sigs_indel |>
  sigrap::sig_contribution_table(type = "ID", outdir = params$outdir) |>
  knitr::kable(format = "html") |>
  kableExtra::kable_styling(c("hover", "striped"), font_size = 12) |>
  kableExtra::scroll_box(height = "400px")
```


## Session Info

```{r session_info, echo=FALSE}
pkgs_of_interest <- c("base", "sigrap", "gpgr", "MutationalPatterns")
si <- gpgr::session_info_kable(pkgs_of_interest)
si$si_pkg
si$si_pl
```
