<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="sigrap">
<title>CHORD â€¢ sigrap</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="CHORD">
<meta property="og:description" content="sigrap">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">sigrap</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/chord.html">CHORD</a>
    <a class="dropdown-item" href="../articles/hrdetect.html">HRDetect</a>
    <a class="dropdown-item" href="../articles/mutationalpatterns.html">MutationalPatterns</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/umccr/sigrap/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>CHORD</h1>
            
            <h4 data-toc-skip class="date">2023-07-12</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/umccr/sigrap/blob/HEAD/vignettes/chord.Rmd" class="external-link"><code>vignettes/chord.Rmd</code></a></small>
      <div class="d-none name"><code>chord.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="va"><a href="https://umccr.github.io/sigrap/">sigrap</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="va"><a href="https://umccr.github.io/gpgr/" class="external-link">gpgr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="va"><a href="https://devtools.r-lib.org/" class="external-link">devtools</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="va"><a href="https://yihui.org/knitr/" class="external-link">knitr</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>CHORD: Classifier of HOmologous Recombination Deficiency (<a href="https://github.com/UMCUGenetics/CHORD" class="external-link uri">https://github.com/UMCUGenetics/CHORD</a>)</p>
<blockquote>
<p>CHORD is a random forest model that uses the relative counts of
somatic mutation contexts to predict homologous recombination deficiency
(HRD). The primary contexts used by CHORD are deletions with flanking
microhomology and 1-100kb structural duplications. Additionally, 1-100kb
structural duplications are used to distinguish BRCA1-type HRD from
BRCA2-type HRD.</p>
</blockquote>
</div>
<div class="section level2">
<h2 id="data">Data<a class="anchor" aria-label="anchor" href="#data"></a>
</h2>
<p>CHORD takes as input somatic VCFs containing SNVs, INDELs and SVs.
The former two can be in the same VCF. It is also able to take
data.frames with those variants as input. Due to the annotation size in
some of our processed Manta VCFs, we prefer to feed it a pre-processed
SV data.frame instead of a VCF, thus we use the
<code><a href="../reference/chord_mantavcf2df.html">sigrap::chord_mantavcf2df</a></code> function.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">snvindel_vcf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/umccrise/snv/somatic-ensemble-PASS.vcf.gz"</span>, package <span class="op">=</span> <span class="st">"gpgr"</span><span class="op">)</span></span>
<span><span class="va">sv_vcf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/umccrise/sv/manta.vcf.gz"</span>, package <span class="op">=</span> <span class="st">"gpgr"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="analysis">Analysis<a class="anchor" aria-label="anchor" href="#analysis"></a>
</h2>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sv_df</span> <span class="op">&lt;-</span> <span class="fu">sigrap</span><span class="fu">::</span><span class="fu"><a href="../reference/chord_mantavcf2df.html">chord_mantavcf2df</a></span><span class="op">(</span><span class="va">sv_vcf</span><span class="op">)</span> <span class="co"># prepare SV VCF as data.frame</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">sigrap</span><span class="fu">::</span><span class="fu"><a href="../reference/chord_run.html">chord_run</a></span><span class="op">(</span></span>
<span>  vcf.snv <span class="op">=</span> <span class="va">snvindel_vcf</span>,</span>
<span>  df.sv <span class="op">=</span> <span class="va">sv_df</span>,</span>
<span>  sv.caller <span class="op">=</span> <span class="st">"manta"</span>,</span>
<span>  <span class="co"># vcf.sv = sv_vcf, # alternative</span></span>
<span>  sample.name <span class="op">=</span> <span class="st">"sample_A"</span>,</span>
<span>  ref.genome <span class="op">=</span> <span class="st">"hg38"</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; #====== Loading variants from vcfs ======#</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ## SNVs</span></span>
<span><span class="co">#&gt; Warning in fun(libname, pkgname): </span></span>
<span><span class="co">#&gt;    No reference genome loaded. Please install and load a BSgenome.</span></span>
<span><span class="co">#&gt;    For example:</span></span>
<span><span class="co">#&gt;       install.packages('BiocManager')</span></span>
<span><span class="co">#&gt;       BiocManager::install('BSgenome.Hsapiens.UCSC.hg19')</span></span>
<span><span class="co">#&gt;       library('BSgenome.Hsapiens.UCSC.hg19')</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;    Then specify the BSgenome to the ref.genome arguemnts to the relevant functions.</span></span>
<span><span class="co">#&gt;    For example:</span></span>
<span><span class="co">#&gt;       extractSigsSnv(..., ref.genome=BSgenome.Hsapiens.UCSC.hg19)</span></span>
<span><span class="co">#&gt; Reading in vcf file...</span></span>
<span><span class="co">#&gt; Converting chrom name style to style in ref.genome...</span></span>
<span><span class="co">#&gt; Loading required package: BiocGenerics</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'BiocGenerics'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:dplyr':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     combine, intersect, setdiff, union</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     IQR, mad, sd, var, xtabs</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     anyDuplicated, aperm, append, as.data.frame, basename, cbind,</span></span>
<span><span class="co">#&gt;     colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,</span></span>
<span><span class="co">#&gt;     get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,</span></span>
<span><span class="co">#&gt;     match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,</span></span>
<span><span class="co">#&gt;     Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,</span></span>
<span><span class="co">#&gt;     table, tapply, union, unique, unsplit, which.max, which.min</span></span>
<span><span class="co">#&gt; Loading required package: S4Vectors</span></span>
<span><span class="co">#&gt; Loading required package: stats4</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'S4Vectors'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:dplyr':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     first, rename</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     expand.grid, I, unname</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'IRanges'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:dplyr':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     collapse, desc, slice</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ## Indels</span></span>
<span><span class="co">#&gt; vcf file is the same for both SNVs and indels. Skipping reading vcf for indels</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ## SVs</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; #====== Counting mutation contexts ======#</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ## Single base substitutions</span></span>
<span><span class="co">#&gt; Loading variants...</span></span>
<span><span class="co">#&gt; Initializing SNV signature output vector...</span></span>
<span><span class="co">#&gt; Removing rows with multiple ALT sequences...</span></span>
<span><span class="co">#&gt; Subsetting for SNVs...</span></span>
<span><span class="co">#&gt; Getting SNV trinucleotide contexts...</span></span>
<span><span class="co">#&gt; Converting trinucleotide contexts to substitution contexts...</span></span>
<span><span class="co">#&gt; Counting substitution context occurrences...</span></span>
<span><span class="co">#&gt; Returning context counts...</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ## Indel contexts (types x lengths)</span></span>
<span><span class="co">#&gt; Loading variants...</span></span>
<span><span class="co">#&gt; Removing rows with multiple ALT sequences...</span></span>
<span><span class="co">#&gt; Determining indel type...</span></span>
<span><span class="co">#&gt; Initializing indel signature output vector...</span></span>
<span><span class="co">#&gt; Determining indel length and sequence...</span></span>
<span><span class="co">#&gt; Determining the start/end positions for the left/right flanks of each indel...</span></span>
<span><span class="co">#&gt; Retrieving flanking sequences...</span></span>
<span><span class="co">#&gt; Calculating the number of copies of the indel sequence are present in the 3' flanking sequence...</span></span>
<span><span class="co">#&gt; Calculating the (max) number of bases that are homologous to the 5'/3' flanking sequence...</span></span>
<span><span class="co">#&gt; Determining indel contexts...</span></span>
<span><span class="co">#&gt; Counting indel context occurrences...</span></span>
<span><span class="co">#&gt; Returning indel context counts...</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ## SV contexts (types x lengths)</span></span>
<span><span class="co">#&gt; Creating SV type/length lookup table...</span></span>
<span><span class="co">#&gt; Counting DEL, DUP, and INV context occurrences...</span></span>
<span><span class="co">#&gt; Counting TRA occurrences...</span></span>
<span><span class="co">#&gt; Returning SV contexts...</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; #====== Exporting output =========#</span></span>
<span><span class="co">#&gt; output.path not specified. Directly returning output</span></span>
<span><span class="co">#&gt; </span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="results">Results<a class="anchor" aria-label="anchor" href="#results"></a>
</h2>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">res</span>, list.len <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 2</span></span>
<span><span class="co">#&gt;  $ contexts  : num [1, 1:145] 123 82 18 91 92 98 21 94 51 55 ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, "dimnames")=List of 2</span></span>
<span><span class="co">#&gt;   .. ..$ : chr "sample_A"</span></span>
<span><span class="co">#&gt;   .. ..$ : chr [1:145] "A[C&gt;A]A" "A[C&gt;A]C" "A[C&gt;A]G" "A[C&gt;A]T" ...</span></span>
<span><span class="co">#&gt;  $ prediction:'data.frame':  1 obs. of  17 variables:</span></span>
<span><span class="co">#&gt;   ..$ sample           : chr "sample_A"</span></span>
<span><span class="co">#&gt;   ..$ p_hrd            : num 0.804</span></span>
<span><span class="co">#&gt;   ..$ hr_status        : chr "HR_deficient"</span></span>
<span><span class="co">#&gt;   ..$ hrd_type         : chr "BRCA2_type"</span></span>
<span><span class="co">#&gt;   .. [list output truncated]</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span>var <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">prediction</span><span class="op">)</span>, value <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">prediction</span><span class="op">[</span><span class="fl">1</span>, <span class="op">]</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span>.name_repair <span class="op">=</span> <span class="st">"check_unique"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span>caption <span class="op">=</span> <span class="st">"CHORD results."</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<caption>CHORD results.</caption>
<thead><tr class="header">
<th align="left">var</th>
<th align="left">value</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">sample</td>
<td align="left">sample_A</td>
</tr>
<tr class="even">
<td align="left">p_hrd</td>
<td align="left">0.804</td>
</tr>
<tr class="odd">
<td align="left">hr_status</td>
<td align="left">HR_deficient</td>
</tr>
<tr class="even">
<td align="left">hrd_type</td>
<td align="left">BRCA2_type</td>
</tr>
<tr class="odd">
<td align="left">p_BRCA1</td>
<td align="left">0.156</td>
</tr>
<tr class="even">
<td align="left">p_BRCA2</td>
<td align="left">0.648</td>
</tr>
<tr class="odd">
<td align="left">remarks_hr_status</td>
<td align="left"></td>
</tr>
<tr class="even">
<td align="left">remarks_hrd_type</td>
<td align="left"></td>
</tr>
<tr class="odd">
<td align="left">p_hrd.5%</td>
<td align="left">0.753</td>
</tr>
<tr class="even">
<td align="left">p_hrd.50%</td>
<td align="left">0.806</td>
</tr>
<tr class="odd">
<td align="left">p_hrd.95%</td>
<td align="left">0.82</td>
</tr>
<tr class="even">
<td align="left">p_BRCA1.5%</td>
<td align="left">0.076</td>
</tr>
<tr class="odd">
<td align="left">p_BRCA1.50%</td>
<td align="left">0.177</td>
</tr>
<tr class="even">
<td align="left">p_BRCA1.95%</td>
<td align="left">0.265</td>
</tr>
<tr class="odd">
<td align="left">p_BRCA2.5%</td>
<td align="left">0.548</td>
</tr>
<tr class="even">
<td align="left">p_BRCA2.50%</td>
<td align="left">0.626</td>
</tr>
<tr class="odd">
<td align="left">p_BRCA2.95%</td>
<td align="left">0.701</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># grab first and last few of the contexts</span></span>
<span><span class="va">col_ind</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span>, <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">contexts</span><span class="op">)</span> <span class="op">-</span> <span class="fl">10</span><span class="op">)</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">contexts</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">res</span><span class="op">$</span><span class="va">contexts</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span>rownames <span class="op">=</span> <span class="st">"context"</span>, .name_repair <span class="op">=</span> <span class="st">"check_unique"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/row_number.html" class="external-link">row_number</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">n</span>, <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/everything.html" class="external-link">everything</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html" class="external-link">slice</a></span><span class="op">(</span><span class="va">col_ind</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span>caption <span class="op">=</span> <span class="st">"Sample of rows from contexts counts."</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<caption>Sample of rows from contexts counts.</caption>
<thead><tr class="header">
<th align="right">n</th>
<th align="left">context</th>
<th align="right">sample_A</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="right">1</td>
<td align="left">A[C&gt;A]A</td>
<td align="right">123</td>
</tr>
<tr class="even">
<td align="right">2</td>
<td align="left">A[C&gt;A]C</td>
<td align="right">82</td>
</tr>
<tr class="odd">
<td align="right">3</td>
<td align="left">A[C&gt;A]G</td>
<td align="right">18</td>
</tr>
<tr class="even">
<td align="right">4</td>
<td align="left">A[C&gt;A]T</td>
<td align="right">91</td>
</tr>
<tr class="odd">
<td align="right">5</td>
<td align="left">C[C&gt;A]A</td>
<td align="right">92</td>
</tr>
<tr class="even">
<td align="right">6</td>
<td align="left">C[C&gt;A]C</td>
<td align="right">98</td>
</tr>
<tr class="odd">
<td align="right">135</td>
<td align="left">DUP_1e04_1e05_bp</td>
<td align="right">4</td>
</tr>
<tr class="even">
<td align="right">136</td>
<td align="left">DUP_1e05_1e06_bp</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="right">137</td>
<td align="left">DUP_1e06_1e07_bp</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="right">138</td>
<td align="left">DUP_1e07_Inf_bp</td>
<td align="right">18</td>
</tr>
<tr class="odd">
<td align="right">139</td>
<td align="left">INV_0e00_1e03_bp</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="right">140</td>
<td align="left">INV_1e03_1e04_bp</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="right">141</td>
<td align="left">INV_1e04_1e05_bp</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="right">142</td>
<td align="left">INV_1e05_1e06_bp</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="right">143</td>
<td align="left">INV_1e06_1e07_bp</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="right">144</td>
<td align="left">INV_1e07_Inf_bp</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="right">145</td>
<td align="left">TRA</td>
<td align="right">77</td>
</tr>
</tbody>
</table>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<table class="table">
<caption>Main packages used in this vignette.</caption>
<thead><tr class="header">
<th align="left">package</th>
<th align="left">version</th>
<th align="left">datestamp</th>
<th align="left">source</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">base</td>
<td align="left">4.2.3</td>
<td align="left">2023-07-12</td>
<td align="left">local</td>
</tr>
<tr class="even">
<td align="left">CHORD</td>
<td align="left">2.0</td>
<td align="left">2023-07-12</td>
<td align="left">local</td>
</tr>
<tr class="odd">
<td align="left">gpgr</td>
<td align="left">1.4.4</td>
<td align="left">2023-07-12</td>
<td align="left">local</td>
</tr>
<tr class="even">
<td align="left">mutSigExtractor</td>
<td align="left">1.25</td>
<td align="left">2023-07-12</td>
<td align="left">local</td>
</tr>
<tr class="odd">
<td align="left">sigrap</td>
<td align="left">0.1.1</td>
<td align="left">2023-07-12</td>
<td align="left">local</td>
</tr>
</tbody>
</table>
<table class="table">
<caption>Platform information.</caption>
<colgroup>
<col width="11%">
<col width="88%">
</colgroup>
<thead><tr class="header">
<th align="left">name</th>
<th align="left">value</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">version</td>
<td align="left">R version 4.2.3 (2023-03-15)</td>
</tr>
<tr class="even">
<td align="left">os</td>
<td align="left">Ubuntu 22.04.2 LTS</td>
</tr>
<tr class="odd">
<td align="left">system</td>
<td align="left">x86_64, linux-gnu</td>
</tr>
<tr class="even">
<td align="left">ui</td>
<td align="left">X11</td>
</tr>
<tr class="odd">
<td align="left">language</td>
<td align="left">en</td>
</tr>
<tr class="even">
<td align="left">collate</td>
<td align="left">C.UTF-8</td>
</tr>
<tr class="odd">
<td align="left">ctype</td>
<td align="left">C.UTF-8</td>
</tr>
<tr class="even">
<td align="left">tz</td>
<td align="left">Etc/UTC</td>
</tr>
<tr class="odd">
<td align="left">date</td>
<td align="left">2023-07-12</td>
</tr>
<tr class="even">
<td align="left">pandoc</td>
<td align="left">2.19.2 @ /usr/share/miniconda/envs/pkgdownenv/bin/ (via
rmarkdown)</td>
</tr>
</tbody>
</table>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by <a href="https://github.com/pdiakumis" class="external-link">Peter Diakumis</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
